orchard_key_components.py

```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env python3
+import sys; assert sys.version_info[0] >= 3, "Python 3 required."
+
+from pyblake2 import blake2b, blake2s
+
+from orchard_generators import NULLIFIER_K_BASE, SPENDING_KEY_BASE, group_hash
+from orchard_pallas import Fp, Scalar, Point
+from orchard_poseidon_hash import poseidon_hash
+from orchard_merkle_tree import MERKLE_DEPTH
+from orchard_commitments import commit_ivk, note_commit
+from utils import leos2bsp, leos2ip, i2leosp
+from tv_output import render_args, render_tv
+
+#
+# Utilities
+#
+
+def to_scalar(buf):
+    return Scalar(leos2ip(buf))
+
+def to_base(buf):
+    return Fp(leos2ip(buf))
+
+
+#
+# PRFs and hashes
+#
+
+def prf_expand(sk: bytes, t: bytes):
+    digest = blake2b(person=b'Zcash_ExpandSeed')
+    digest.update(sk)
+    digest.update(t)
+    return digest.digest()
+
+def diversify_hash(d):
+    P = group_hash(b'z.cash:Orchard-gd', d)
+    if P == Point.identity():
+        P = group_hash(b'z.cash:Orchard-gd', b'')
+    return P
+
+def prf_nf_orchard(nk, rho):
+    return poseidon_hash(nk, rho)
+
+def derive_nullifier(nk, rho: Fp, psi: Fp, cm):
+    scalar = prf_nf_orchard(nk, rho) + psi  # addition mod p
+    point = NULLIFIER_K_BASE * Scalar(scalar.s) + cm
+    return point.extract()
+
+#
+# Key components
+#
+
+class SpendingKey:
+    def __init__(self, data):
+        self.data = data
+
+        self.ask  = to_scalar(prf_expand(self.data, b'\x06'))
+        self.nk   = to_base(prf_expand(self.data, b'\x07'))
+        self.rivk = to_scalar(prf_expand(self.data, b'\x08'))
+        if self.ask == Scalar.ZERO:
+            raise ValueError("invalid spending key")
+
+        self.akP = SPENDING_KEY_BASE * self.ask
+        if bytes(self.akP)[-1] & 0x80 != 0:
+            self.ask = -self.ask
+
+        self.ak = self.akP.extract()
+        if commit_ivk(self.rivk, self.ak, self.nk) is None:
+            raise ValueError("invalid spending key")
```

Comment:
Because `sinsemilla_short_commit` uses `checked_incomplete_add` which asserts that exceptional cases do not occur, we will never hit `None` here:
```suggestion
        assert commit_ivk(self.rivk, self.ak, self.nk) is not None
```